loudcloud-dev/ee_calculator
# feat: leave model and form

## Changelog

### Features

#### Model
- added `Leave` model
    - `employee_id` - associated with employee
    - `approver_id` - associated with admin user
    - `start_date`
    - `end_date`
    - `day_count`
    - `leave_type`
    - `reason` - optional
- added custom validation methods, as a fallback if the front-end form restrictions are bypassed

#### Controller
- added basic controller for `Leave`
    - only actions for now are index, new, and create 
    - require authentication for any action, since only users should be able to access leaves

#### View
- form for creating new leave requests
    - connected to stimulus controller that handles front-end validations and other methods
- basic leave index page, to be updated on a future PR

## How to check
- log in to an employee account
- navigate to Leaves Tracker -> File Leave Request
- fill up the form
    - when selecting a start date, verify that user is not allowed to select an end date that falls before that date
    - when selecting an end date, verify that user is not allowed to select a start date that falls after that date
    - verify that user cannot input a day count lower than 1 or higher than 15
    - when both start and end dates are selected, verify that day count automatically updates to count the number of days, not counting weekdays
    - when day count is manually changed, verify that end date automatically updates to that amount of days after the selected start date
- submit the form
    - verify that the user is prompted if they leave required fields empty (all fields are required except reason)
    - verify that if user somehow bypasses the day count limit (by selecting start and end dates that amount to more than 15 days), the user will be prompted about it
